File: apps/mobile/ios/HermesMobile/State/EventReplayGuard.swift
import Foundation

@MainActor
public final class EventReplayGuard {
    private struct Key: Hashable {
        let storedSessionID: String
        let eventType: String
        let stableIdentity: String
    }

    private let capacity: Int
    private var activeProfileScope: String?
    private var sequence: UInt64 = 0
    private var seen: [Key: UInt64] = [:]

    public init(capacity: Int = 1024) {
        precondition(capacity > 0, "Replay capacity must be positive")
        self.capacity = capacity
    }

    public func activate(profileScope: String) {
        let normalized = profileScope.trimmingCharacters(in: .whitespacesAndNewlines)
        precondition(!normalized.isEmpty, "Profile scope must not be empty")
        if activeProfileScope != normalized {
            seen.removeAll(keepingCapacity: true)
            sequence = 0
            activeProfileScope = normalized
        }
    }

    public func shouldConsume(
        event: GatewayEvent,
        storedSessionID: String,
        profileScope: String
    ) -> Bool {
        let normalizedProfile = profileScope.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalizedProfile.isEmpty,
              activeProfileScope == normalizedProfile else { return false }
        let stored = storedSessionID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !stored.isEmpty, let identity = stableIdentity(for: event) else { return true }

        sequence &+= 1
        let key = Key(
            storedSessionID: stored,
            eventType: event.type.rawValue,
            stableIdentity: identity
        )
        if seen[key] != nil {
            seen[key] = sequence
            return false
        }
        seen[key] = sequence
        evictIfNeeded()
        return true
    }

    public func clear() {
        seen.removeAll(keepingCapacity: true)
        sequence = 0
        activeProfileScope = nil
    }

    private func evictIfNeeded() {
        while seen.count > capacity {
            guard let oldest = seen.min(by: { $0.value < $1.value })?.key else { return }
            seen.removeValue(forKey: oldest)
        }
    }

    private func stableIdentity(for event: GatewayEvent) -> String? {
        if let value = primitiveIdentity(event.payload?.object?["event_id"]) {
            return "event_id:\(value)"
        }
        if let value = primitiveIdentity(event.payload?.object?["eventId"]) {
            return "event_id:\(value)"
        }
        if let value = primitiveIdentity(event.payload?.object?["sequence"]) {
            return "sequence:\(value)"
        }

        let fields: [String]
        switch event.type {
        case .messageStart, .messageComplete:
            fields = ["message_id"]
        case .toolStart, .toolComplete:
            fields = ["tool_id", "tool_call_id"]
        case .approvalRequest, .clarifyRequest, .clarifyExpire,
             .secretRequest, .secretExpire, .sudoRequest, .sudoExpire:
            fields = ["request_id", "id"]
        default:
            fields = []
        }
        for field in fields {
            if let value = primitiveIdentity(event.payload?.object?[field]) {
                return "\(field):\(value)"
            }
        }
        return nil
    }

    private func primitiveIdentity(_ value: JSONValue?) -> String? {
        guard let value else { return nil }
        switch value {
        case .string(let string):
            let normalized = string.trimmingCharacters(in: .whitespacesAndNewlines)
            return normalized.isEmpty ? nil : normalized
        case .number(let number):
            guard number.isFinite,
                  number.rounded() == number,
                  let integer = Int(exactly: number) else { return nil }
            return String(integer)
        default:
            return nil
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/snowzlmbot/hermes/mobile/feature/EventReplayGuard.kt
package com.snowzlmbot.hermes.mobile.feature

import com.snowzlmbot.hermes.mobile.core.GatewayEvent
import com.snowzlmbot.hermes.mobile.core.GatewayEventType
import kotlinx.serialization.json.JsonPrimitive

internal class EventReplayGuard(private val capacity: Int = 1024) {
  init {
    require(capacity > 0) { "Replay capacity must be positive" }
  }

  private data class Key(
    val storedSessionId: String,
    val eventType: String,
    val stableIdentity: String,
  )

  private var activeProfileScope: String? = null
  private val seen = LinkedHashMap<Key, Unit>(capacity, 0.75f, true)

  @Synchronized
  fun activate(profileScope: String) {
    val normalized = profileScope.trim()
    require(normalized.isNotEmpty()) { "Profile scope must not be empty" }
    if (activeProfileScope != normalized) {
      seen.clear()
      activeProfileScope = normalized
    }
  }

  @Synchronized
  fun accept(profileScope: String, storedSessionId: String, event: GatewayEvent): Boolean {
    val normalizedProfile = profileScope.trim()
    if (normalizedProfile.isEmpty()) return false
    if (activeProfileScope == null) activeProfileScope = normalizedProfile
    if (activeProfileScope != normalizedProfile) return false
    val normalizedStoredId = storedSessionId.trim()
    if (normalizedStoredId.isEmpty()) return true
    val identity = stableIdentity(event) ?: return true
    val key = Key(normalizedStoredId, event.wireType, identity)
    if (seen[key] != null) return false
    seen[key] = Unit
    while (seen.size > capacity) {
      val eldest = seen.entries.iterator()
      eldest.next()
      eldest.remove()
    }
    return true
  }

  @Synchronized
  fun clear() {
    seen.clear()
    activeProfileScope = null
  }

  private fun stableIdentity(event: GatewayEvent): String? {
    primitiveIdentity(event.payload["event_id"])?.let { return "event_id:$it" }
    primitiveIdentity(event.payload["eventId"])?.let { return "event_id:$it" }
    primitiveIdentity(event.payload["sequence"])?.let { return "sequence:$it" }
    return fallbackFields(event.type)
      .firstNotNullOfOrNull { field -> primitiveIdentity(event.payload[field])?.let { "$field:$it" } }
  }

  private fun primitiveIdentity(value: kotlinx.serialization.json.JsonElement?): String? {
    val primitive = value as? JsonPrimitive ?: return null
    if (primitive.isString) return primitive.content.trim().takeIf(String::isNotEmpty)
    return primitive.content.toLongOrNull()?.toString()
  }

  private fun fallbackFields(type: GatewayEventType): List<String> = when (type) {
    GatewayEventType.MESSAGE_START,
    GatewayEventType.MESSAGE_COMPLETE -> listOf("message_id")
    GatewayEventType.TOOL_START,
    GatewayEventType.TOOL_COMPLETE -> listOf("tool_id", "tool_call_id")
    GatewayEventType.APPROVAL_REQUEST,
    GatewayEventType.CLARIFY_REQUEST,
    GatewayEventType.CLARIFY_EXPIRE,
    GatewayEventType.SECRET_REQUEST,
    GatewayEventType.SECRET_EXPIRE,
    GatewayEventType.SUDO_REQUEST,
    GatewayEventType.SUDO_EXPIRE -> listOf("request_id", "id")
    else -> emptyList()
  }
}
